File: benchmark_read_tabulated.m
%% Code to benchmark reading tabulated data
%% Settings
% Set paths, relative to this script
workDir     = fileparts(mfilename('fullpath'));
inDir       = fullfile(workDir, 'samples');
resultsDir  = fullfile(workDir, 'results');
if ~exist(resultsDir, 'dir')
    mkdir(resultsDir);
end

fname1  = 'ABCDlike_tabulated_DK40_33794_70';
fname2  = 'ABCDlike_tabulated_GP_32945_335';

%% First benchmark - ABCDlike_tabulated_DK40_33794_70
fcsv     = @() readtable(fullfile(inDir,   [fname1, '.csv']));
fParquet = @() parquetread(fullfile(inDir, [fname1, '.parquet']));

% Get robust timing using timeit
tReadCSV_DK40_33794_70     = timeit(fcsv);
tReadParquet_DK40_33794_70 = timeit(fParquet);

%% Second benchmark - ABCDlike_tabulated_GP_32945_335
fcsv     = @() readtable(fullfile(inDir,   [fname2, '.csv']));
fParquet = @() parquetread(fullfile(inDir, [fname2, '.parquet']));

% Get robust timing using timeit
tReadCSV_GP_32945_335     = timeit(fcsv);
tReadParquet_GP_32945_335 = timeit(fParquet);

%% Make a results table
results      = cell(2, 4);

% File names
results{1,1} = fname1;
results{2,1} = fname2;

% Performance of readtable
results{1,2} = tReadCSV_DK40_33794_70;
results{2,2} = tReadCSV_GP_32945_335;

% Performance of parquetread
results{1,3} = tReadParquet_DK40_33794_70;
results{2,3} = tReadParquet_GP_32945_335;

% Speed-up factor
results{1,4} = tReadCSV_DK40_33794_70 ./ tReadParquet_DK40_33794_70;
results{2,4} = tReadCSV_GP_32945_335  ./ tReadParquet_GP_32945_335;

% Show results
disp(['Time taken: read CSV DK40_33794_70: ',     num2str(tReadCSV_DK40_33794_70)]);
disp(['Time taken: read parquet DK40_33794_70: ', num2str(tReadParquet_DK40_33794_70)]);

disp(['Time taken: read CSV GP_32945_335: ',     num2str(tReadCSV_GP_32945_335)]);
disp(['Time taken: read parquet GP_32945_335: ', num2str(tReadParquet_GP_32945_335)]);

% Make a table
results = cell2table(results, 'VariableNames', {'FileName', 'tReadCSV', 'tReadParquet', 'Parquet_SpeedUp'});

%% Save results
clear f*
save(fullfile(resultsDir, 'benchmarks_readTabulated_MATLAB.mat'));

File: benchmark_read_tabulated.R
# Get data.table and arrow packages
if (!require(data.table, quietly = TRUE))
{
  install.packages("data.table")
}
library("data.table")
if (!require(arrow, quietly = TRUE))
{
  install.packages("arrow")
}
library("arrow")
if (!require("this.path", quietly = TRUE))
{
  install.packages("this.path")
}
library(this.path)

# Get the path of this script
workDir    <- this.path::this.dir()
inDir      <- file.path(workDir, "samples")
resultsDir <- file.path(workDir, "results")
dir.create(resultsDir, showWarnings = FALSE, recursive = TRUE)

# Benchmark reading of the files
fname1         <- "ABCDlike_tabulated_DK40_33794_70"
fname2         <- "ABCDlike_tabulated_GP_32945_335"
numRepeats     <- 10

# Initialize
tCSV_readCSV_DK40_33794_70 <- vector("numeric", length = numRepeats)
tCSV_fread_DK40_33794_70   <- vector("numeric", length = numRepeats)
tCSV_arrow_DK40_33794_70   <- vector("numeric", length = numRepeats)
tParquet_DK40_33794_70     <- vector("numeric", length = numRepeats)

tCSV_readCSV_GP_32945_335 <- vector("numeric", length = numRepeats)
tCSV_fread_GP_32945_335   <- vector("numeric", length = numRepeats)
tCSV_arrow_GP_32945_335   <- vector("numeric", length = numRepeats)
tParquet_GP_32945_335     <- vector("numeric", length = numRepeats)

toRead_CSV     <- file.path(inDir, paste0(fname1, ".csv"))
toRead_parquet <- file.path(inDir, paste0(fname1, ".parquet"))
for (rep in 1:numRepeats)
{
  # Read using R's base read.csv
  tCSV_readCSV_DK40_33794_70[rep] <- system.time(read.csv(toRead_CSV))["elapsed"]
  
  # Read using data table's fread
  tCSV_fread_DK40_33794_70[rep] <- system.time(fread(toRead_CSV))["elapsed"]
  
  # Read using arrow's read csv
  tCSV_arrow_DK40_33794_70[rep] <- system.time(read_csv_arrow(toRead_CSV))["elapsed"]
  
  # Finally, read using parquet format
  tParquet_DK40_33794_70[rep] <- system.time(read_parquet(toRead_parquet))["elapsed"]
}

toRead_CSV     <- file.path(inDir, paste0(fname2, ".csv"))
toRead_parquet <- file.path(inDir, paste0(fname2, ".parquet"))
for (rep in 1:numRepeats)
{
  # Read using R's base read.csv
  tCSV_readCSV_GP_32945_335[rep] <- system.time(read.csv(toRead_CSV))["elapsed"]
  
  # Read using data table's fread
  tCSV_fread_GP_32945_335[rep] <- system.time(fread(toRead_CSV))["elapsed"]
  
  # Read using arrow's read csv
  tCSV_arrow_GP_32945_335[rep] <- system.time(read_csv_arrow(toRead_CSV))["elapsed"]
  
  # Finally, read using parquet format
  tParquet_GP_32945_335[rep] <- system.time(read_parquet(toRead_parquet))["elapsed"]
}

save.image(file=file.path(resultsDir, "benchmarks_readTabulated_R.rdata"))
